Xenpcid format2

NastyaVicodin opened the pull request.

Commit:
tools/libs/light: Add vchan client to libxl PCI

Add possibility to send commands from libxl PCI and execute them on server
side - xenpcid daemon.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/libs/light: Add vchan client to libxl PCI

Add possibility to send commands from libxl PCI and execute them on server
side - xenpcid daemon.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add xenpcid daemon

Add draft version of xenpcid server (based on vchan-node2), which can
receive messages from the client.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add xenpcid daemon

Add draft version of xenpcid server (based on vchan-node2), which can
receive messages from the client.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+        if ( sscanf(domains[i], "%d", &d) != 1 )
+            continue;
+        if ( xs_path_exists(gc, GCSPRINTF(XENPCID_XS_PATH, d)) > 0 )
+        {
```
formatting: `{` should be on the same line as `if`

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+    libxl_domid domid = DOMID_INVALID;
+
+    domains = libxl__xs_directory(gc, XBT_NULL, "/local/domain", &n);
+    if ( !n )
```
formatting: no spaces around `!n`. 

Unfortunately, tools use coding style that differs from hypervisor :(

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+    for (i = 0; i < n; i++) {
+        int d;
+
+        if ( sscanf(domains[i], "%d", &d) != 1 )
```
formatting: no spaces inside braces. This applies to all other places with such mistake (I will not point all of them)

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+        return ERROR_NOTFOUND;
+    }
+
+    state->xs_path = strdup(GCSPRINTF(XENPCID_XS_PATH, state->domid));
```
why do you need to `strdup` path there and manually free it later? You already have GC context there.

NastyaVicodin replied:
Agree

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+
+    hand = libxl_yajl_gen_alloc(NULL);
+
+    if ( !hand ) {
```
No need for curly braces around single statement.

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+    }
+
+#if HAVE_YAJL_V2
+    /* Disable beautify for data sent to QEMU */
```
QEMU?

NastyaVicodin replied:
This code is based on qmp_prepare_cmd function from libxl_qmp. I think this comment can be deleted

lorc replied:
Well, yes. Either remove the comment at all, or remove the "QEMU" part

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+
+static int vchan_get_next_msg(libxl__gc *gc, struct vchan_state *state,
+                              libxl__json_object **o_r)
+    /*
```
I mentioned this earlier: place this comment **before** the function declaration

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+{
+    size_t len;
+    char *end = NULL;
+    const char eom[] = XENPCID_END_OF_MESSAGE;
```
why do you need separate variable with that constant? You can use it right away.

NastyaVicodin replied:
Agree

lorc commented on tools/include/xenpcid.h:
```diff
@@ -0,0 +1,49 @@
+
+#define XENPCID_XS_PATH             "/local/domain/%d/data/pcid-vchan"
+
+#define XENPCID_END_OF_MESSAGE      "\r\n"
```
Question about protocol: how do you ensure that `\r\n` will not happen in the message text itself?

NastyaVicodin replied:
Now all sent messages are essentially constants and file paths. Of course, in theory, these symbols can be the text of messages, but in practice I do not see a situation in which this is possible.

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,358 @@
+            LOGE(ERROR, "vchan write failed");
+            return ERROR_FAIL;
+        }
+        if ( ret > 0 )
```
Looking at the function logic, this `if` is not needed. At this point ret is `>=0`, so you can alter `cmd` and `ret` in any case.

NastyaVicodin replied:
Agree

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -428,9 +773,11 @@ static void pci_info_xs_remove(libxl__gc *gc, libxl_device_pci *pci,
 {
     GC_INIT(ctx);
-    libxl_device_pci *pcis = NULL, *new;
+    libxl_pci_bdf *pcibdfs = NULL, *new;
```
looks like this change does not belongs to this patch

NastyaVicodin replied:
Agree

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -446,6 +793,67 @@ libxl_device_pci *libxl_device_pci_assignable_list(libxl_ctx *ctx, int *num)
         goto out;
     }
 
+    while((de = readdir(dir))) {
```
looks like this change does not belongs to this patch

NastyaVicodin replied:
Agree

lorc commented on tools/include/xenpcid.h:
```diff
@@ -36,6 +36,8 @@
 
 #define XENPCID_PCIBACK_DRIVER      "pciback_driver"
 
+#include <stdbool.h>
```
Why do we need to add this, when there are no users in the header file?

NastyaVicodin replied:
Agree

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -114,6 +114,14 @@ static int vchan_init_client(libxl__gc *gc, struct vchan_state *state)
     return 0;
 }
 
+/*
```
I believe this comment should be added in the previous patch ('Add vchan client to libxl PCI")

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/Makefile:
```diff
@@ -0,0 +1,32 @@
+include $(XEN_ROOT)/tools/Rules.mk
+
+CFLAGS += -Werror
+CFLAGS += $(CFLAGS_libxenstore) $(CFLAGS_libxenvchan) $(CFLAGS_libxengnttab) $(CFLAGS_libxenevtchn) $(CFLAGS_libxenlight)
```
Do we really need `libxengnttab` and `libxenevtchn`?

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/pcid.h:
```diff
@@ -0,0 +1,100 @@
+#define PCI_BDF_XSPATH         "%04x-%02x-%02x-%01x"
+#define PCI_BDF                "%04x:%02x:%02x.%01x"
+
+enum pcid__json_node_type {
```
Those constants are already defined in "libxl_internal.h". As we can't use this header from this project, we should move the JSON definitions to separate header file and include it from both the projects.

I am not asking to do this right now, but you need to add the comment with TODO there.

lorc replied:
On other hand I can see that you are calling libxl function from this module. So probably you can include that libxl header as well?

NastyaVicodin replied:
Added TODO comment

lorc commented on tools/xenpcid/pcid.h:
```diff
@@ -0,0 +1,100 @@
+    char *val;
+};
+
+struct flexarray {
```
The same about `flexarray`. It is already implemented in liblight.

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+/*
```
We need the proper header with license text there

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    return ptr;
+}
+
+static void vchan_wr(char *reply)
```
`reply` is not the best name. IMO, `message` or `msg` will be better.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    int ret, len;
+
+    len = strlen(reply);
+    ret = libxenvchan_write(ctrl, reply, len);
```
There is a possibility, that this function will not write all the data in one call. You need at least to check that `ret == len`. But better approach is to implement write loop like one you used in the previous patch.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    }
+
+#if HAVE_YAJL_V2
+    /* Disable beautify for data sent to QEMU */
```
QEMU?

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+        goto out;
+    }
+
+    ret = pcid_zalloc(sizeof((int)len));
```
You are allocating 4 bytes and then...

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    }
+
+    ret = pcid_zalloc(sizeof((int)len));
+    sprintf(ret, "%*.*s" XENPCID_END_OF_MESSAGE,
```
... `sprintf` there giant buffer?  This is likely a bug

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    return o != NULL && o->type == JSON_ARRAY;
+}
+
+static int pcid__json_object_append_to(libxl_ctx *ctx,
```
what is `libxl_ctx` and where it is used?

NastyaVicodin replied:
Removed as it's used nowhere

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    if (pcid__json_object_append_to(ctx->ctx, obj, ctx))
+        return 1;
+
+    return 1;
```
On the previous `if` you also have `return 1`. Something is wrong there.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    return flexarray_set(array, array->count, ptr);
+}
+
+static struct flexarray *flexarray_make(libxl_ctx *ctx, int size, int autogrow)
```
Why do we need `ctx`?

NastyaVicodin replied:
Removed

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    obj->u.string = t;
+
+    if (pcid__json_object_append_to(ctx->ctx, obj, ctx))
+        return 1;
```
`return 0`, probably. Please check.

NastyaVicodin replied:
Yes

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    memset(&yajl_ctx, 0, sizeof (yajl_ctx));
+    yajl_ctx.ctx = ctx;
+
+    if (yajl_ctx.hand == NULL) {
```
You just called `memset(&yajl_ctx, 0, sizeof (yajl_ctx))`. Of course there will be NULL.

NastyaVicodin replied:
Check removed

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+
+static int vchan_get_next_msg(libxl_ctx *ctx, struct vchan_state *state,
+                              struct pcid__json_object **o_r)
+    /* Find a JSON object and store it in o_r.
```
Place the comment before the function declaration

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+
+out:
+    str = yajl_get_error(yajl_ctx.hand, 1, (const unsigned char*)s, strlen(s));
+    yajl_free_error(yajl_ctx.hand, str);
```
Well, you might print the error if you got it.

Or what is the purpose of this?

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+        return ERROR_PROTOCOL_ERROR_PCID;
+    }
+    state->rx_buf_used = len;
+    state->rx_buf_used -= len;
```
Is this a fancy way to write `state->rx_buf_used = 0`? Or it is some mistake?

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    fprintf(stderr, "parsing %luB: '%.*s'\n", len, (int)len,
+         state->rx_buf);
+
+    if (len) {
```
Looks like this function mostly operates on `state`. Is this a right place to touch the global `outbuf`? And why it is `outbuf`  and not `inbuf`?

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    return 0;
+}
+
+static int vchan_read_message(libxl_ctx *ctx,
```
`read_message` is not the best name, taking into account that this function calls `handle_message`

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+    ret = vchan_read_message(ctx, state, &result);
+
+    reply = vchan_prepare_cmd(result, args, 0);
+    if ( !reply ){
```
formatting is wrong. Also, this will break the channel, because client is waiting for at least some response.

You need to prepare static string with the JSON-formatted error response, that can be sent if you can't create a correct response

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,637 @@
+        FD_SET(libxenvchan_fd, &rfds);
+        ret = select(libxenvchan_fd + 1, &rfds, &wfds, NULL, NULL);
+        if (ret < 0) {
+            exit(1);
```
Some error message is welcomed there.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/pcid.h:
```diff
@@ -0,0 +1,100 @@
+    JSON_ANY     = 255 /* this is a mask of all values above, adjust as needed */
+};
+
+struct list_head {
```
Looks like you are not using this structure in this patch. So it should be moved to the patch, where it is used for the first time.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -134,6 +134,53 @@ static char *vchan_prepare_cmd(struct pcid__json_object *result,
 
+static int handle_ls_command(char *dir_name, struct list_head **result)
+{
+    struct list_head *dirs = NULL, *head = NULL, *prev =NULL;
```
Do you need a linked list if you already have flexarray?

NastyaVicodin replied:
I thought that for this type of information it is easier to use a list than to constantly make reallocations using flexarray

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -134,6 +134,53 @@ static char *vchan_prepare_cmd(struct pcid__json_object *result,
+        goto out;
+    }
+
+    if (NULL == dir) {
```
`dir == NULL`

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -229,13 +276,35 @@ static int vchan_handle_message(libxl_ctx *ctx,
     command_obj = pcid__json_map_get(XENPCID_MSG_EXECUTE, *resp, JSON_ANY);
     command_name = command_obj->u.string;
 
+    if (strcmp(XENPCID_CMD_LIST, command_name) == 0) {
```
`strcmp(command_name, XENPCID_CMD_LIST)`

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -134,7 +134,6 @@ static struct vchan_state *vchan_get_instance(libxl__gc *gc)
     ret = vchan_init_client(gc, state);
     if ( ret )
         state = NULL;
-
```
Spurious change

NastyaVicodin replied:
Done

lorc replied:
But it is still there :)

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -845,10 +846,35 @@ libxl_device_pci *libxl_device_pci_assignable_list(libxl_ctx *ctx, int *num)
+
+        new = realloc(pcis, ((*num) + 1) * sizeof(*new));
+        if (NULL == new)
+            continue;
```
Well, this is an error. You should fail there. Also `new == NULL`.

NastyaVicodin replied:
Done

lorc replied:
I can't see any changes to this

NastyaVicodin replied:
Sorry, previously I accidentally did it only for #else part. Now it's done for both.

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -252,8 +252,11 @@ static int vchan_handle_message(libxl__gc *gc,
     {
     case LIBXL__PCID_MESSAGE_TYPE_RETURN:
         *result = libxl__json_map_get(XENPCID_MSG_RETURN, resp, JSON_ARRAY);
+        if ( !(*result) )
```
What is the logic behind this `if`?

NastyaVicodin replied:
If the result is not found in the form of an array, then you need to check other types of the returned result - string, map, etc.

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -129,6 +141,7 @@ static char *vchan_prepare_cmd(struct pcid__json_object *result,
     ret = pcid_zalloc(sizeof((int)len));
     sprintf(ret, "%*.*s" XENPCID_END_OF_MESSAGE,
             (int)len, (int)len, buf);
+    fprintf(stderr, "ret = %s\n", ret);
```
Looks like an spurious change.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -271,14 +284,36 @@ static struct pcid__json_object *pcid__json_map_get(const char *key,
+    }
+
+    rc = write(fd, pci_info, strlen(pci_info));
+    /* Annoying to have two if's, but we need the errno */
```
I can't see where `errno` is used

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -293,8 +328,18 @@ static int vchan_handle_message(libxl_ctx *ctx,
             goto out;
 
-        (*resp)->type = JSON_STRING;
+        (*resp)->type = JSON_ARRAY;
```
Why this change?

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -306,12 +310,33 @@ static int handle_write_cmd(char *sysfs_path, char *pci_info)
+                sysfs_path);
+        return -1;
+    }
+    read_items = fscanf(f, "0x%llx\n", &result);
```
So, this is not a generic "read" command, but a command that reads a hexadecimal value?

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -306,12 +310,33 @@ static int handle_write_cmd(char *sysfs_path, char *pci_info)
+    FILE *f = fopen(sysfs_path, "r");
+
+    if (!f) {
+        fprintf(stderr, "PCI device %s does not have vendor attribute\n",
```
Why "vendor"? Looks like liblx uses this to read vendor, device and class attributes.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -365,6 +383,26 @@ static int vchan_handle_message(libxl_ctx *ctx,
+            free(pci_info->u.string);
+        } else
+            full_path = pci_info->u.string;
+        ret = handle_lstat_cmd(full_path);
```
So, actually this is not a "lstat" command, but just something like "is_exists", right?
Because you actually never return stat structure anywhere.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -162,6 +162,8 @@ static int handle_ls_command(char *dir_name, struct list_head **result)
 
     if (strcmp(XENPCID_PCIBACK_DRIVER, dir_name) == 0) {
         dir = opendir(SYSFS_PCIBACK_DRIVER);
+    } else if (strcmp(XENPCID_PCI_DEV, dir_name) == 0){
```
Formatting is wrong

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -130,14 +154,14 @@ static char *vchan_prepare_cmd(struct pcid__json_object *result,
                 free(result->u.string);
-            }
-            else
+            } else
```
Spirous change?

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -2234,7 +2257,9 @@ static int libxl__device_pci_reset(libxl__gc *gc, unsigned int domain, unsigned
     } else {
         LOGED(ERROR, domain, "Failed to access reset path %s", reset);
     }
+
```
Spurious change

NastyaVicodin replied:
Done

lorc replied:
But it is still there :)

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -2664,6 +2689,82 @@ static void do_pci_remove(libxl__egc *egc, pci_remove_state *prs)
+            }
+        }
+
+skip1:
```
`skip1` is very uninformative name for a label

NastyaVicodin replied:
It was taken from this function's code. I've decided not to change the names of labels

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -2664,6 +2689,82 @@ static void do_pci_remove(libxl__egc *egc, pci_remove_state *prs)
+        if ( !vchan )
+            goto out_fail;
+        libxl__qmp_param_add_string(gc, &args, XENPCID_CMD_PCI_INFO, sysfs_path);
+        result = vchan_send_command(gc, vchan, XENPCID_CMD_READ_RESOURCES, args);
```
This commit is about "reset" and "remove" commands. So why this command is called in this commit?

NastyaVicodin replied:
"Remove" functions consists of several stages, which includes a command XENPCID_CMD_READ_RESOURCES. Therefore, when creating a patch that adds the ability to do "remove" using xenpcid, I decided to replace all parts at once.

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,1056 @@
+    command_name = command_obj->u.string;
+
+    if (strcmp(XENPCID_CMD_LIST, command_name) == 0) {
+        args = pcid__json_map_get(XENPCID_MSG_FIELD_ARGS, *resp, JSON_MAP);
```
`pcid__json_map_get()` there and elsewhere below can return `NULL`. You need to check for this before trying using the value.

lorc replied:
Also, it will be move all this (and for other `if`branges)  logic  to `handle_*_function`.  

`vchan_handle_message` will be much easier to read after this.

NastyaVicodin replied:
Done

lorc reviewed (changes requested):
0. It is good to look thru patches before submitting them for
      review. At least you need to check for spurious changes and make
      sure that changes in the patch correspond to the commit message.

   1. Please take a look at how to format commit descriptions. Try
      `git log --one-line tools'. You will see that all commits have
      prefix like "libxl/arm: " or something like that. You need 
      to do the same

   2. Code formatting. Unfortunately, tools use different coding
      style. You need to ensure that coding style of your changes
      corresponds to coding style of the file.

   3. Code reuse. Looks like pcid uses some structures and functions
      (like flexarrays, json handling, etc) from libxl. It is bad to
      duplicate code. Better to reuse existing one.

   4. Commit "Optimize executing libxl PCI commands to sysfs via
      xenpcid daemon" does many different things for one commit. I
      think, you need to create separate commit that tidies code in
      libxl_pci, without ever calling xenpcid. And then add separate
      commit that adds support for xenpcid.

   5. Commit "Add free memory functions to avoid memory leaks" should
      not exist at all. All changes from this commit should go to the
      commits that adds corresponding functionality.

Commit:
tools/xenpcid: Add "ls" command processing to xenpcid daemon

Messaging format is based on json style. Xenpcid receives "ls" command
from client and sends a reply:
1. success case - array of directory's names
2. error case - message with "error" in return

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "ls" command processing to xenpcid daemon

Messaging format is based on json style. Xenpcid receives "ls" command
from client and sends a reply:
1. success case - array of directory's names
2. error case - message with "error" in return

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "write" command to xenpcid and libxl PCI

Add functions for handling "write" command on xenpcid side. Libxl PCI side
receives reply:
1. success - string in "return" message
2. fail - error type of reply message

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "write" command to xenpcid and libxl PCI

Add functions for handling "write" command on xenpcid side. Libxl PCI side
receives reply:
1. success - string in "return" message
2. fail - error type of reply message

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "read" command to xenpcid and libxl PCI

Libxl PCI can use xenpcid server to read integer value from the given path.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "read" command to xenpcid and libxl PCI

Libxl PCI can use xenpcid server to read integer value from the given path.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Make Libxl PCI get values from xenpcid instead of libxl side

pci_multifunction_check needs to use "ls" and "lstat" commands to get
information from sysfs.
Add "is_exists" command processing to xenpcid daemon to make possible
pci_multifunction_check read directories and check lstat by using xenpcid.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Make Libxl PCI get values from xenpcid instead of libxl side

pci_multifunction_check needs to use "ls" and "lstat" commands to get
information from sysfs.
Add "is_exists" command processing to xenpcid daemon to make possible
pci_multifunction_check read directories and check lstat by using xenpcid.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "reset" and "remove" PCI dev commands to xenpcid daemon and libxl PCI

Add "reset" command processing to xenpcid daemon.
Make possible sending "reset"/"remove" PCI device requests from libxl PCI
to xenpcid daemon instead of using sysfs on libxl side.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "reset" and "remove" PCI dev commands to xenpcid daemon and libxl PCI

Add "reset" command processing to xenpcid daemon.
Make possible sending "reset"/"remove" PCI device requests from libxl PCI
to xenpcid daemon instead of using sysfs on libxl side.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Daemonize xenpcid server

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Daemonize xenpcid server

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -445,6 +783,40 @@ libxl_device_pci *libxl_device_pci_assignable_list(libxl_ctx *ctx, int *num)
+    GC_INIT(ctx);
+    *num = 0;
+
+#ifdef XENPCID_SERVER_USE
```
`XENPCID_SERVER_USE` sounds a bit unnatural. Maybe `USE_XENPCID_SERVER` or just `XENPCID_SERVER` ?

NastyaVicodin replied:
Agree

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -428,6 +764,8 @@ static void pci_info_xs_remove(libxl__gc *gc, libxl_device_pci *pci,
 
 libxl_device_pci *libxl_device_pci_assignable_list(libxl_ctx *ctx, int *num)
+{
+#if 0
```
Why are you disabling some portions of code? Better to remove it all

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,349 @@
+{
+    libxl__pcid_message_type type;
+    libxl__json_map_node *node = NULL;
+    int i = 0;
```
Don't need to initialize `i` with `0` is it is initialized in the `for` loop

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,349 @@
+                                          state->xs_path);
+    if (!state->ctrl) {
+        LOGE(ERROR, "Couldn't intialize vchan client");
+        free(state->xs_path);
```
`xs_path` buffer was allocated with `GCSPRINTF`. I believe, it should be destroyed automatically when `gc` state is freed. If you want to destroy it earlier, you need use gc-compatible function, or you will get double free

NastyaVicodin replied:
Done

lorc commented on tools/libs/light/libxl_pci.c:
```diff
@@ -18,13 +18,349 @@
+        free(state->xs_path);
+        return ERROR_FAIL;
+    }
+    state->select_fd = libxenvchan_fd_for_select(state->ctrl);
```
I dig code a bit and looks like `libxenvchan_fd_for_select` can fail and return -1. I believe, you need to check return value for errors.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(1);
+    }
+    memset(ptr, 0, size);
```
`calloc` already zeroes memory:

> The  calloc()  function  allocates memory for an array of nmemb elements of size bytes each and returns a pointer to the allocated memory.  The memory is set to zero. 

(from man page)

NastyaVicodin replied:
Agree

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+        count = map->count;
+        while (count--) {
+            struct pcid__json_map_node *node;
+            node = map->data[count];
```
empty line is missing between variable declaration and code

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+    newsize = array->size + extents;
+    array->data = realloc(array->data, newsize * sizeof(*(array->data)));
+    if (!(array->data)) {
+        fprintf(stderr, "Memory reallocation for data in array failed\n");
```
There you will lost previously allocated buffer:

>  If realloc() fails, the original block is left untouched; it is not freed or moved.

You need to save value of `array->data` somewhere before calling `realloc`.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+{
+    if (idx >= array->size) {
+        int newsize;
+        if (!array->autogrow)
```
Missing empty line between variable declaration and code

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+    obj->u.string = NULL;
+
+    if (type == JSON_MAP || type == JSON_ARRAY) {
+        struct flexarray *array = flexarray_make(1, 1);
```
Missing empty line between variable declaration and code

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+    yajl_ctx.hand = libxl__yajl_alloc(&callbacks, NULL, &yajl_ctx);
+
+    status = yajl_parse(yajl_ctx.hand, (const unsigned char *)s, strlen(s));
+    if (status != yajl_status_ok){
```
Missing space `) {`

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+    }
+
+    status = yajl_complete_parse(yajl_ctx.hand);
+    if (status != yajl_status_ok){
```
Missing space `) {`

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+
+    status = yajl_parse(yajl_ctx.hand, (const unsigned char *)s, strlen(s));
+    if (status != yajl_status_ok){
+        fprintf(stderr, "yajl_parse error\n");
```
It will be nice to print error code.

NastyaVicodin replied:
Done

lorc commented on tools/xenpcid/xenpcid.c:
```diff
@@ -0,0 +1,675 @@
+
+    status = yajl_complete_parse(yajl_ctx.hand);
+    if (status != yajl_status_ok){
+        fprintf(stderr, "yajl_complete_parse error\n");
```
It will be nice to print error code.

NastyaVicodin replied:
Done

lorc commented on tools/hotplug/Linux/systemd/xen-pcid.service.in:
```diff
@@ -0,0 +1,14 @@
+Type=forking
+EnvironmentFile=-@CONFIG_DIR@/@CONFIG_LEAF_DIR@/xencommons
+ExecStart=@sbindir@/xenpcid
+KillSignal=USR1
```
According to this, server should be stopped with `SIGUSR1` signal, but I can't see a signal handler anywhere in the code.

lorc reviewed (changes requested):
1. It would be great to control deamonization  with command line options, not with build options.

2. During daemonization it is advisable to close `stderr` and `stdout`. And write all logs to log file.

Commit:
tools/libs/light: Reading assignable PCI devices using xenpcid server

Xenpcid server transmits information about devices assigned to the PCI
driver by sending reply with directory's names. Libxl PCI processes reply
message instead of opening directory in libxl side.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/libs/light: Reading assignable PCI devices using xenpcid server

Xenpcid server transmits information about devices assigned to the PCI
driver by sending reply with directory's names. Libxl PCI processes reply
message instead of opening directory in libxl side.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
tools/xenpcid: Add "unbind" and "read resources" commands to libxl PCI and xenpcid

Send requests to xenpcid from libxl PCI to prevent the libxl PCI from
reading / writing from / to the local sysfs directly.

"Unbind" command returns driver's path or "nolstat" result in success and
NULL if writing to PCI path failed.

"Read resources" command returns an array of PCI resources - start, end and
flags in success, NULL if getting resources from PCI path failed.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

Commit:
xen/ipmmu-vmsa-plat.c: Fix compile error due to missing include

Fix the following compile error due to missing mm.h include:

ipmmu-vmsa-plat.c:122:19: error: implicit declaration of function 'ioremap_nocache' [-Werror=implicit-function-declaration]

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Reviewed-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>

Commit:
Revert "[HACK][DO NOT MERGE] Hack is assignable"

This reverts commit c243af51247d7156d1df6b64fbf81046c1fca375.

Commit:
Revert "[HACK][DO NOT MERGE] Hack is assignable"

This reverts commit c243af51247d7156d1df6b64fbf81046c1fca375.

Commit:
tools/xenpcid: Add "unbind" and "read resources" commands to libxl PCI and xenpcid

Send requests to xenpcid from libxl PCI to prevent the libxl PCI from
reading / writing from / to the local sysfs directly.

"Unbind" command returns driver's path or "nolstat" result in success and
NULL if writing to PCI path failed.

"Read resources" command returns an array of PCI resources - start, end and
flags in success, NULL if getting resources from PCI path failed.

Signed-off-by: Anastasiia Lukianenko <anastasiia_lukianenko@epam.com>

NastyaVicodin committed:
Merge branch 'pci_phase2_pre' into xenpcid_format2
